server/src/main/java/io/crate/metadata/doc/DocIndexMetadata.java

```diff
@@ -280,6 +280,14 @@ private void addGeoReference(Integer position,
         }
     }
 
+    @Nullable
+    private static Symbol castIfNeeded(@Nullable Symbol defaultExpression, DataType<?> type) {
+        if (defaultExpression != null && !type.equals(defaultExpression.valueType())) {
```

Comment:
`.cast` already optimizes this:

https://github.com/crate/crate/blob/d5edce334d1a898e729bf6b9bfdcea64a106ca36/server/src/main/java/io/crate/expression/symbol/Symbol.java#L74-L75

A `castIfNonNull(symbol, type)` would be enough

And I think we should add an assertion to the Reference constructors to disallow creating inconsistent References. (Or even do the cast there, both seems okay to me)